Slack cannot receive TAG_RETENTION webhook

How can we help you?

As the invalid payload of tag retention, so currently the slack cannot receive the webhook notification successfully from harbor.

```
2023-03-25T08:59:27Z [INFO] [/jobservice/job/impl/notification/slack_job.go:81]: start to run slack job
2023-03-25T08:59:27Z [INFO] [/jobservice/job/impl/notification/slack_job.go:120]: send request to remote endpoint, body: {
	"blocks": [
		{
            "type": "section",
			"text": {
				"type": "mrkdwn",
				"text": "*Harbor webhook events*"
			}
        },
        {
            "type": "section",
			"text": {
				"type": "mrkdwn",
				"text": "*event_type:* TAG_RETENTION"
			}
        },
        {
            "type": "section",
			"text": {
				"type": "mrkdwn",
				"text": "*occur_at:* <!date^1679734643^{date} at {time}|February 18th, 2014 at 6:39 AM PST>"
			}
        },
        {	"type": "section",
			"text": {
				"type": "mrkdwn",
				"text": "*operator:* MANUAL"
			}
		},
        {	"type": "section",
			"text": {
				"type": "mrkdwn",
				"text": "*event_data:*"
			}
		},
		{	"type": "section",
			"text": {
				"type": "mrkdwn",
				"text": "```{
	\"retention\": {
		\"total\": 1,
		\"retained\": 0,
		\"harbor_hostname\": \"localhost\",
		\"project_name\": \"library\",
		\"retention_policy_id\": 1,
		\"retention_rule\": [
			{
				\"template\": \"always\",
				\"tag_selectors\": [
					{
						\"kind\": \"doublestar\",
						\"decoration\": \"matches\",
						\"pattern\": \"11\",
						\"extras\": \"{\\"untagged\\":true}\"
					}
				],
				\"scope_selectors\": {
					\"repository\": [
						{
							\"kind\": \"doublestar\",
							\"decoration\": \"repoMatches\",
							\"pattern\": \"busybox\",
							\"extras\": \"\"
						}
					]
				}
			}
		],
		\"result\": \"SUCCESS\",
		\"deleted_artifact\": [
			{
				\"type\": \"image\",
				\"status\": \"SUCCESS\",
				\"name_tag\": \"busybox:latest\"
			}
		]
	}
}```"
			}
		}
    ]}
2023-03-25T08:59:27Z [ERROR] [/jobservice/job/impl/notification/slack_job.go:85]: exit slack job, error: abnormal response code: 400, body: invalid_payload
```

--- a/src/pkg/notifier/handler/notification/slack_handler.go
+++ b/src/pkg/notifier/handler/notification/slack_handler.go
@@ -121,7 +121,7 @@ func (s *SlackHandler) convert(payLoad *model.Payload) (string, error) {
 	if err != nil {
 		return "", fmt.Errorf("marshal from eventData %v failed: %v", payLoad.EventData, err)
 	}
-	data["EventData"] = "```" + strings.Replace(string(eventData), `"`, `\"`, -1) + "```"
+	data["EventData"] = "```" + escapeEventData(string(eventData)) + "```"
 
 	st, _ := template.New("slack").Parse(SlackBodyTemplate)
 	var slackBuf bytes.Buffer
@@ -129,4 +129,12 @@ func (s *SlackHandler) convert(payLoad *model.Payload) (string, error) {
 		return "", fmt.Errorf("%v", err)
 	}
 	return slackBuf.String(), nil
+}
+
+func escapeEventData(str string) string {
+	// escape " to \"
+	str = strings.Replace(str, `"`, `\"`, -1)
+	// escape \\" to \\\"
+	str = strings.Replace(str, `\\"`, `\\\"`, -1)
+	return str
 }

--- a/src/pkg/notifier/handler/notification/slack_handler_test.go
+++ b/src/pkg/notifier/handler/notification/slack_handler_test.go
@@ -8,7 +8,6 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
-	"github.com/goharbor/harbor/src/common/dao"
 	"github.com/goharbor/harbor/src/pkg/notification"
 	policy_model "github.com/goharbor/harbor/src/pkg/notification/policy/model"
 	"github.com/goharbor/harbor/src/pkg/notifier/event"
@@ -16,7 +15,7 @@ import (
 )
 
 func TestSlackHandler_Handle(t *testing.T) {
-	dao.PrepareTestForPostgresSQL()
+
 	hookMgr := notification.HookManager
 	defer func() {
 		notification.HookManager = hookMgr
@@ -106,4 +105,33 @@ func TestSlackHandler_IsStateful(t *testing.T) {
 func TestSlackHandler_Name(t *testing.T) {
 	handler := &SlackHandler{}
 	assert.Equal(t, "Slack", handler.Name())
+}
+
+func Test_escapeEventData(t *testing.T) {
+	type args struct {
+		str string
+	}
+	tests := []struct {
+		name string
+		args args
+		want string
+	}{
+		{
+			name: `escape "`,
+			args: args{str: `{"foo":"bar"}`},
+			want: `{\"foo\":\"bar\"}`,
+		},
+		{
+			name: `escape \\"`,
+			args: args{str: `{\"foo\":\"bar\"}`},
+			want: `{\\\"foo\\\":\\\"bar\\\"}`,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := escapeEventData(tt.args.str); got != tt.want {
+				t.Errorf("escapeEventData() = %v, want %v", got, tt.want)
+			}
+		})
+	}
 }

